design/airgap-existing-cluster.md

```diff
@@ -0,0 +1,110 @@
+1. The user downloads the kotsadm archive containing kots container images
+1. The user downloads the airgap bundle of the application
+1. The user copies the kots plugin, the kotsadm archive and the application airgap bundle onto a medium (eg a USB stick) that can be transfered into the airgapped network.
+
+From the airgap workstation:
+1. The user will install the kots plugin 
+1. The user will copy the container images in the kotsadm archive to the local registry using the kots kubectl plugin
+1. The user will install kotsadm to the cluster using the kots plugin via `kots install` passing local registry information which will be used to generate the kotsadm yaml and image pull secrets (if neccessary)
+1. The user will install thte application and license using the existing airgap installation process 
+
+## Detailed Design
+
+The CI process, on tag, will create the kotsadm archive and include it in the GitHub Release.
+The filename will be `kots_airgap_<version>.tar.gz` and this file contains .tar file exports of all kotsadm images.
+The files SHOULD be stored in the archive in the same format as application airgap files are stored.
+
+Add a `kots admin-console push-images` to the kots cli to retag and push all kotsadm images to an authenticated registry that is accessible from the cluster, using the desired push credentials.
+This new command is used to copy all of the KOTS Admin Console images from one location to another.
+
+Example:
+
+```
+kubectl kots admin-console push-images ./kotsadm-1.16.1.tar.gz registry.somebigbank.com/my-app \
+  --registry-username push-user \
+  --registry-password abcdef
+```
+
+`--registry-username` and `--registry-password` are both optional.
+If not present, attempt to authenticate using the already authenticated local docker credentials.
+Stretch: if still not authenticated, prompt for username/password, use these to push but do not save locally.
+Stretch: if username is specified and password is not, prompt for password to use, do not save locally.
+
+KOTS will then load the images from the file in the first parameter, retag them to the local registry and push. 
+All images will retain the same image name and tag, just the registry, namespace will change.
+
+At this point, the images are in the registry, but not running in the cluster yet.
+A workflow could scan these images before continuing.
+
+Next, add a flag to `kots install` to collect / rewrite local registry endpoint/namespace into kots manifests, using the desired pull credentials.
+This will assume that the images are already in the destination.
+
+For example:
+
+```
+kubectl kots install --kotsadm-registry registry.somebigbank.com/my-app \
+  --registry-username pull-user \
+  --registry-password fedcba \
+  app-name
+```
+
+If a `registry-username` and `registry-password` are provided, an `imagePullSecret` will be created and deployed with the registry credentials and added to all kotsadm pods.
```

Comment:
I think we also need to consider the case where no registry-username/password are provided, but the user wants to direct kots to add an already-existing imagePullSecret to all workloads

Reply:
@dexhorthy Hmm... how likely is that? Can we start without it? We briefly discussed using advanced secret management like short-lived Vault credentials but decided it was out of scope. If the existing imagePullSecret is a simple username and password, then having to replicate it isn't horrible at least, although it's not DRY and may miss regular rotating if there's some existing mechanism to rotate creds. I think we can cut that for now.